staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -185,98 +202,147 @@ func (q *delayingType) AddAfter(item interface{}, duration time.Duration) {
 
-		case <-nextReadyAt:
-			// continue the loop, which will add ready items
+		case <-w.nextReadyAt:
+			w.processWaitingItems()
 
-		case waitEntry := <-q.waitingForAddCh:
-			if waitEntry.readyAt.After(q.clock.Now()) {
-				insert(waitingForQueue, waitingEntryByData, waitEntry)
-			} else {
-				q.Add(waitEntry.data)
-			}
+		case waitEntry := <-w.q.waitingForAddCh:
+			w.processArrivingEntry(waitEntry)
 
 			drained := false
 			for !drained {
 				select {
-				case waitEntry := <-q.waitingForAddCh:
-					if waitEntry.readyAt.After(q.clock.Now()) {
-						insert(waitingForQueue, waitingEntryByData, waitEntry)
-					} else {
-						q.Add(waitEntry.data)
-					}
+				case waitEntry := <-w.q.waitingForAddCh:
+					w.processArrivingEntry(waitEntry)
 				default:
 					drained = true
 				}
 			}
 		}
+		// Recalculate the nextReadyTime after every signal is processed.
+		w.calculateNextReadyTime()
+	}
+}
+
+// processWaitingItems adds waiting items that are ready and returns a timer to first item in the queue that will be ready next.
+func (w *waitingLoop) processWaitingItems() {
+	now := w.q.clock.Now()
+
+	// Add ready entries
+	for w.waitingForQueue.Len() > 0 {
+		entry := w.waitingForQueue.Peek().(*waitFor)
+		if entry.readyAt.After(now) {
+			break
+		}
+
+		entry = heap.Pop(w.waitingForQueue).(*waitFor)
+		w.q.Interface.Add(entry.data)
+		delete(w.waitingEntryByData, entry.data)
+	}
+}
+
+// calculateNextReadyTime creates a new timer and signal channel for the first ready time.
+func (w *waitingLoop) calculateNextReadyTime() {
+	now := w.q.clock.Now()
+
+	// Set up a wait for the first item's readyAt (if one exists)
+	if w.waitingForQueue.Len() > 0 {
+		if w.nextReadyAtTimer != nil {
+			w.nextReadyAtTimer.Stop()
+		}
+		entry := w.waitingForQueue.Peek().(*waitFor)
+		w.nextReadyAtTimer = w.q.clock.NewTimer(entry.readyAt.Sub(now))
+		w.nextReadyAt = w.nextReadyAtTimer.C()
+	}
+}
+
+// processArrivingEntry decides whether to accept or drop an entry based on its waitFor settings and state of the queue.
+// It implements the following behaviours:
+//   - An entry is dropped if it is already actively queued (much like Add() on the Interface)
+//   - Delayed items are inserted into the waiting queue.
+//   - Immediate items will be dropped when the dropIfWaiting option is set and they are already waiting.
+//   - Immediate items will pre-empt/remove waiting items when dropIfWaiting is false (default)
+func (w *waitingLoop) processArrivingEntry(entry *waitFor) {
+	// If the entry already exists in the active queue don't insert again.
+	if w.q.Interface.IsQueued(entry.data) {
+		return
+	}
+
+	// delayed add
+	if entry.readyAt.After(w.q.clock.Now()) {
+		// inform the metrics of the retry (moved from AddAfter())
+		w.q.metrics.retry()
+		w.insert(entry)
+		return
+	}
+
+	// immediate add
+	existing, exists := w.waitingEntryByData[entry.data]
+	if exists {
+		// drop if the incoming entry has dropIfWaitingSet
+		if entry.dropIfWaiting {
```

Comment:
hm I think I expect this to be checked in the "delayed add" case, too?

Reply:
Ok, agreed.  I didn't do this originally because AddAfter is naturally deduped anyway, but I totally agree that it feels inconsistent to not extend the entry.dropIfWaiting behaviour to AddAfter.  To expose this we'll need another method like "AddIfNotWaiting" - so I've added "AddIfNotWaitingAfter" which performs exactly as AddAfter does but with entry.dropIfWaiting.